File: examples/python/atlas_examples.py
#!/usr/bin/env python3
"""Dependency-free 2b2tAtlas public API examples."""

from __future__ import annotations

import json
import os
import sys
import urllib.parse
import urllib.request

API = os.environ.get("ATLAS_API_BASE_URL", "https://api.blackportal.cloud").rstrip("/")


def get(path: str, params: dict[str, object] | None = None):
    query = "?" + urllib.parse.urlencode(params) if params else ""
    request = urllib.request.Request(
        API + path + query,
        headers={
            "Accept": "application/json",
            "User-Agent": "2b2tAtlas-API-Examples/1.0",
        },
    )
    with urllib.request.urlopen(request, timeout=30) as response:
        return json.load(response)


def normalize(value: str) -> str:
    return " ".join(value.casefold().split())


def main() -> int:
    search = " ".join(sys.argv[1:]) or "Mu Megabase"
    locations = get("/api/locations")
    needle = normalize(search)
    location = next((item for item in locations if normalize(item["name"]) == needle), None)
    location = location or next((item for item in locations if needle in normalize(item["name"])), None)
    if location is None:
        print(f"No location matched: {search}", file=sys.stderr)
        return 1

    print(
        f'{location["name"]} [{location["dimensionName"]}] '
        f'{location["x"]}, {location["y"]}, {location["z"]}'
    )
    print(location["interactiveUrl"])

    location_id = location["rowid"]
    warps = get("/api/warps", {"locationId": location_id, "limit": 1000})
    renders = get("/api/renders", {"locationId": location_id, "limit": 1000})
    print("warps:", [f'/warp {item["name"]}' for item in warps])
    print("renders:", [(item.get("worldDownloadDate"), item["apiUrl"]) for item in renders])

    if location.get("groups"):
        group = get(f'/api/groups/{location["groups"][0]["groupId"]}')
        print(f'{group["name"]}: {group["locationCount"]} builds, {group["highwayCount"]} highways')
        for build in group["locations"][:5]:
            print(f'  build: {build["name"]} ({build["role"]}) -> {build["locationInteractiveUrl"]}')

    highways = get("/api/highways")
    for highway in (item for item in highways if "+Z" in item["name"]):
        print(f'highway: {highway["name"]} | dimension {highway["dimension"]} | {highway["apiUrl"]}')

    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: examples/javascript/atlas-examples.mjs
import http from "node:http";
import https from "node:https";

const API = (process.env.ATLAS_API_BASE_URL || "https://api.blackportal.cloud").replace(/\/$/, "");
const search = process.argv.slice(2).join(" ") || "Mu Megabase";

function get(path) {
  return new Promise((resolve, reject) => {
    const transport = API.startsWith("https:") ? https : http;
    const request = transport.get(`${API}${path}`, {
      headers: {
        Accept: "application/json",
        "User-Agent": "2b2tAtlas-Public-API/1.0",
      },
      timeout: 30_000,
    }, response => {
      let body = "";
      response.setEncoding("utf8");
      response.on("data", chunk => { body += chunk; });
      response.on("end", () => {
        if (!response.statusCode || response.statusCode < 200 || response.statusCode >= 300) {
          reject(new Error(`${response.statusCode} ${response.statusMessage}: ${path}`));
          return;
        }
        try {
          resolve(JSON.parse(body));
        } catch (error) {
          reject(new Error(`Invalid JSON from ${path}: ${error.message}`));
        }
      });
    });
    request.on("timeout", () => request.destroy(new Error(`Timed out: ${path}`)));
    request.on("error", reject);
  });
}

const normalize = value => value.normalize("NFKC").trim().toLocaleLowerCase("en-US");

async function main() {
  const locations = await get("/api/locations");
  const needle = normalize(search);
  const location = locations.find(item => normalize(item.name) === needle)
    || locations.find(item => normalize(item.name).includes(needle));

  if (!location) throw new Error(`No location matched: ${search}`);

  console.log(`${location.name} [${location.dimensionName}] ${location.x}, ${location.y}, ${location.z}`);
  console.log(location.interactiveUrl);

  const [warps, renders] = await Promise.all([
    get(`/api/warps?locationId=${location.rowid}&limit=1000`),
    get(`/api/renders?locationId=${location.rowid}&limit=1000`),
  ]);
  console.log("warps:", warps.map(item => `/warp ${item.name}`));
  console.log("renders:", renders.map(item => ({
    date: item.worldDownloadDate,
    apiUrl: item.apiUrl,
    sourceWdl: item.worldDownloadUrl || (item.archiveWarp && item.archiveWarp.worldDownloadUrl) || null,
    blueMapUrl: item.blueMapUrl || null,
    blueMapProfileVersion: item.blueMapProfileVersion || null,
  })));

  if (location.groups && location.groups.length) {
    const group = await get(`/api/groups/${location.groups[0].groupId}`);
    console.log(`${group.name}: ${group.locationCount} builds, ${group.highwayCount} highways`);
    console.table(group.locations.slice(0, 5).map(item => ({
      build: item.name,
      role: item.role,
      url: item.locationInteractiveUrl,
    })));
  }

  const highways = await get("/api/highways");
  console.table(highways.filter(item => item.name.includes("+Z")).slice(0, 8).map(item => ({
    name: item.name,
    dimension: item.dimension,
    points: item.points.length,
    apiUrl: item.apiUrl,
  })));
}

main().catch(error => {
  console.error(error.message);
  process.exitCode = 1;
});
